asv/plugins/benchmarks/asvmemray.py

```diff
@@ -0,0 +1,299 @@
+                itertools.product(*self._params),
+                param_idx, param_idx + 1)
+        except ValueError:
+            raise ValueError(
+                f"Invalid benchmark parameter permutation index: {param_idx!r}")
+
+    def insert_param(self, param):
+        """
+        Insert a parameter at the front of the parameter list.
+        """
+        self._current_params = tuple([param] + list(self._current_params))
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__} {self.name}>'
+
+    def check(self, root):
+        # Check call syntax (number of arguments only...)
+        ok = True
+
+        if self._params:
+            self.set_param_idx(0)
+
+        min_num_args = len(self._current_params)
+        max_num_args = min_num_args
+
+        if self.setup_cache_key is not None:
+            ok = ok and check_num_args(root, self.name + ": setup_cache",
+                                       self._setup_cache, 0)
+            max_num_args += 1
+
+        for setup in self._setups:
+            ok = ok and check_num_args(root, self.name + ": setup",
+                                       setup, min_num_args, max_num_args)
+
+        ok = ok and check_num_args(root, self.name + ": call",
+                                   self.func, min_num_args, max_num_args)
+
+        for teardown in self._teardowns:
+            ok = ok and check_num_args(root, self.name + ": teardown",
+                                       teardown, min_num_args, max_num_args)
+
+        return ok
+
+    def do_setup(self):
+        try:
+            for setup in self._setups:
+                setup(*self._current_params)
+        except NotImplementedError as e:
+            # allow skipping test
+            print(f"asv: skipped: {e !r} ")
+            return True
+        return False
+
+    def redo_setup(self):
+        if not self._redo_setup_next:
+            self._redo_setup_next = True
+            return
+        self.do_teardown()
+        self.do_setup()
+
+    def do_teardown(self):
+        for teardown in self._teardowns:
+            teardown(*self._current_params)
+
+    def do_setup_cache(self):
+        if self._setup_cache is not None:
+            return self._setup_cache()
+
+    def do_run(self):
+        return self.run(*self._current_params)
+
+    def do_profile(self, filename=None):
+        def method_caller():
+            run(*params)  # noqa:F821 undefined name  see #1020 Bug: run() function is not defined
+
+        if profile is None:
+            raise RuntimeError("cProfile could not be imported")
+
+        if filename is not None:
+            if hasattr(method_caller, 'func_code'):
+                code = method_caller.func_code
+            else:
+                code = method_caller.__code__
+
+            self.redo_setup()
+
+            profile.runctx(
+                code, {'run': self.func, 'params': self._current_params},
+                {}, filename)
```

Comment:
This is the minimal amount of code needed from `benchmark.py` in order to write "external" plugins. @mattip perhaps this could be spun out into a separate pure-python package? If it has no dependencies other than simple `python` like `benchmark.py` it shouldn't cause any additional overhead in installations.

Reply:
> this could be spun out into a separate pure-python package?

Do you mean create a `asv-memray` package and put it on PyPI, so it can be pip-installed? Sounds good to me, but maybe there are additional opinions: @pv, @datapythonista ?

Reply:
Yup I could do that, but also I was thinking of a bit of a broader refactor:
- Spin out the pure-python part (highlighted) into something like `asv-core` or `asv-bench`
- Then use `asv-core` in both `asv` and `asv-memray`

The goal is to separate out the "pure python" part of the benchmark into its own package which can be installed without causing dependency issues with other packages, while still letting plugins (and `asv`) have other dependencies.